peerconnection_js.go

```diff
@@ -0,0 +1,543 @@
+}
+
+// PendingRemoteDescription represents a remote description that is in the
+// process of being negotiated, complete with any remote candidates that
+// have been supplied via AddICECandidate() since the offer or answer was
+// created. If the PeerConnection is in the stable state, the value is
+// null.
+func (pc *PeerConnection) PendingRemoteDescription() *SessionDescription {
+	desc := pc.underlying.Get("pendingRemoteDescription")
+	return valueToSessionDescription(desc)
+}
+
+// SignalingState returns the signaling state of the PeerConnection instance.
+func (pc *PeerConnection) SignalingState() SignalingState {
+	rawState := pc.underlying.Get("signalingState").String()
+	return newSignalingState(rawState)
+}
+
+// ICEGatheringState attribute the ICE gathering state of the PeerConnection
+// instance.
+func (pc *PeerConnection) ICEGatheringState() ICEGatheringState {
+	rawState := pc.underlying.Get("iceGatheringState").String()
+	return newICEGatheringState(rawState)
+}
+
+// ConnectionState attribute the connection state of the PeerConnection
+// instance.
+func (pc *PeerConnection) ConnectionState() PeerConnectionState {
+	rawState := pc.underlying.Get("connectionState").String()
+	return newPeerConnectionState(rawState)
+}
+
+// Converts a Configuration to js.Value so it can be passed
+// through to the JavaScript WebRTC API. Any zero values are converted to
+// js.Undefined(), which will result in the default value being used.
+func configurationToValue(configuration Configuration) js.Value {
+	return js.ValueOf(map[string]interface{}{
+		"iceServers":           iceServersToValue(configuration.ICEServers),
+		"iceTransportPolicy":   stringEnumToValueOrUndefined(configuration.ICETransportPolicy.String()),
+		"bundlePolicy":         stringEnumToValueOrUndefined(configuration.BundlePolicy.String()),
+		"rtcpMuxPolicy":        stringEnumToValueOrUndefined(configuration.RTCPMuxPolicy.String()),
+		"peerIdentity":         stringToValueOrUndefined(configuration.PeerIdentity),
+		"iceCandidatePoolSize": uint8ToValueOrUndefined(configuration.ICECandidatePoolSize),
+
+		// TODO(albrow): Docs for RTCCertificate are underspecified.
+		// https://developer.mozilla.org/en-US/docs/Web/API/RTCCertificate How
+		// should we handle this?
+		// "certificates": configuration.Certificates,
+	})
+}
+
+func iceServersToValue(iceServers []ICEServer) js.Value {
+	if len(iceServers) == 0 {
+		return js.Undefined()
+	}
+	maps := make([]interface{}, len(iceServers))
+	for i, server := range iceServers {
+		maps[i] = iceServerToValue(server)
+	}
+	return js.ValueOf(maps)
+}
+
+func iceServerToValue(server ICEServer) js.Value {
+	return js.ValueOf(map[string]interface{}{
+		"urls":           stringsToValue(server.URLs), // required
+		"username":       stringToValueOrUndefined(server.Username),
+		"credential":     interfaceToValueOrUndefined(server.Credential),
+		"credentialType": stringToValueOrUndefined(server.CredentialType.String()),
```

Comment:
Apparently Firefox doesn't allow passing `unknown` to `credentialType`:
```
panic: JavaScript error: 'credentialType' member of RTCIceServer 'unknown' is not a valid value for enumeration RTCIceCredentialType. wasm_exec.js:85:11
wasm_exec.js:85:11
goroutine 1 [running]: wasm_exec.js:85:11
syscall/js.Value.New(0x7ff800030000001c, 0xc06be48, 0x1, 0x1, 0x0) wasm_exec.js:85:11
	syscall/js/js.go:330 +0x9 wasm_exec.js:85:11
github.com/pions/webrtc.NewPeerConnection(0xc02a4c0, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...) wasm_exec.js:85:11
	github.com/pions/webrtc/peerconnection_js.go:50 +0xf wasm_exec.js:85:11
main.main() wasm_exec.js:85:11
	github.com/pions/webrtc/examples/data-channels-wasm/browser/main.go:20 +0x8
```
Maybe there is a difference between passing undefined and not creating the attribute?

Reply:
I think we can ticket this one out if you prefer to land first.

Reply:
No I know how to address this. It'll be a quick fix.

Reply:
@backkem I think this is fixed now. Mind trying again?

You were on the right track and all I had to do was convert `"unknown"` to `js.Undefined`.

Reply:
Cool, I can confirm this works 👍